### DISCUSSION TITLE
v4

### DISCUSSION THREAD

**Discussion by ayushmanchhabra** [nwutils/nw-builder#949] [Announcements]
### v4.x

- [ ] get binaries
  - [x] verify checksum (#593)
  - [x] cache binaries
  - [x] override cache
  - [ ] add support for linux arm (#494)
  - [x] add support for mac arm (#546)
- [ ] `build` mode
  - [x] place app code in `package.nw`
  - [x] platform specific rc config
  - [x] rename macos helpers (#966)
  - [ ] source code protection (#757)
  - [x] update package.json for production usage (#842)
  - [ ] native node modules
    - [x] gyp (#758)
    - [ ] nan (#970)
    - [ ] ffi
    - [ ] napi
  - [ ] code sign binaries (#763)
  - [x] compress outDir (#887)
- [ ] `package` mode RFC (#611)
  - [ ] NSIS Installer (#885)
  - [ ] AppImage Installer (#886 )
  - [ ] DMG Installer (#801)
  - [ ] #898
  - [ ] Snap store compatible
  - [ ] Mac Store compatible

**Comment by sysrage** [comment]
> delete `outDir` dir post zip

This one should be optional.

I'm very interested to see what you find with "chmod binaries" and "code sign binaries".

For the chmod issue in my scripts, I ended up requiring builds to be done in Linux to preserve file permissions from the NW.js tarball as well as dependencies. Otherwise, it was going to be a nightmare to determine which files needed permission changes and to get those permission changes inside a tarball generated in Windows.

I really need to implement code signing, to hopefully avoid the Windows Defender popups. Hoping you make good progress here!

Finally, one more suggested change/enhancement is to implement `rcedit` to modify the Windows NW.js executable (icon and metadata). I suspect something similar is also needed for macOS packages.

P.S. All of these changes sound like a 4.1, not a 5.0. 😉

**Comment by ayushmanchhabra** [comment]
> For the chmod issue in my scripts, I ended up requiring builds to be done in Linux to preserve file permissions from the NW.js tarball as well as dependencies. Otherwise, it was going to be a nightmare to determine which files needed permission changes and to get those permission changes inside a tarball generated in Windows.

In Windows, `chmod` only changes the write permission. Possible solution is to pass in an octal value which represents that if platform is Windows.

> Finally, one more suggested change/enhancement is to implement rcedit to modify the Windows NW.js executable (icon and metadata).

✅ 

> I suspect something similar is also needed for macOS packages.

Yup, it involves updating the `Info.plist` files. First need a way to build and test run the demo app in CI so one would not need to have a Mac to debug NW.js Mac apps.
